Quic natsupp

martinduke opened the pull request:
Resolves #146

britram commented on draft-ietf-quic-manageability.md:
```diff
@@ -905,6 +905,119 @@ need to be flow based.  However, 5-tuple (plus eventually connection ID if
+attractive, and how NAT use of CID can create serious problems for the
+endpoints.
+
+The conclusion of this section is that NATs should retain their existing
```
I'd maybe move this up, or summarize it in the first para. This section should be very clear about the "don't do it"ness of messing with CIDs.

britram commented on draft-ietf-quic-manageability.md:
```diff
@@ -905,6 +905,119 @@ need to be flow based.  However, 5-tuple (plus eventually connection ID if
+Unfortunately, the change of IP address or port is an important signal to QUIC
+endpoints. It requires a review of path-dependent variables like congestion
+control parameters. It can also signify various attacks that mislead one
+endpoint about the best peer address for the connection (see section 9 of 
```
```suggestion
endpoint about the best peer address for the connection (see section 9 of
```

britram replied:
\o/ trailing whitespace

britram commented on draft-ietf-quic-manageability.md:
```diff
@@ -905,6 +905,119 @@ need to be flow based.  However, 5-tuple (plus eventually connection ID if
+internet, which would thwart various important protocol properties.
+
+It is possible, in principle, to encode the client's identity in a connection ID
+using {{QUIC_LB}} and explicit coordination with the NAT. However, QUIC-LB makes assumptions about endpoint mobility and common configuration in server
```
```suggestion
using {{QUIC_LB}} and explicit coordination with the NAT.
However, QUIC-LB makes assumptions about endpoint mobility and
common configuration in server
```

britram replied:
80-char limit

britram commented on draft-ietf-quic-manageability.md:
```diff
@@ -905,6 +905,119 @@ need to be flow based.  However, 5-tuple (plus eventually connection ID if
+
+### The Lure of the Connection ID
+
+There are a few reasons that CID-aware NATs could seemingly appear attractive.
```
editorial, like the comment above: please end this with a snarky conclusion that they are all wrong. :)

martinduke replied:
Made obsolete by Mirja's suggestion

martinduke replied:
overcome by events

britram commented on draft-ietf-quic-manageability.md:
```diff
@@ -905,6 +905,119 @@ need to be flow based.  However, 5-tuple (plus eventually connection ID if
+
+NATs sometimes hit an operational limit where they exhaust available public IP
+addresses and ports, and must evict flows from their address/port mapping. CIDs
+offer a way to multiplex many connections over a single address and port.
```
```suggestion
might appear to offer a way to multiplex many connections over a single
address and port.
```

britram commented on draft-ietf-quic-manageability.md:
```diff
@@ -905,6 +905,119 @@ need to be flow based.  However, 5-tuple (plus eventually connection ID if
+the connection will suddenly break.
+
+While mid-connection failure in some cases may seem superior to rejecting QUIC
+outright, HTTP/3 over QUIC falls back to TCP. This is preferable to a
```
also note that applicability recommends fallback for non H3 protocols.

britram reviewed (changes requested):
LGTM modulo editorial nits / requests

mirjak commented on draft-ietf-quic-manageability.md:
```diff
@@ -905,6 +905,119 @@ need to be flow based.  However, 5-tuple (plus eventually connection ID if
+of source IP address means that a NAT rebinding or address migration will
+deliver packets to the wrong server. For the reasons described above, routers
+and switches will not have access to negotiated CIDs. This is a particular
+problem for low-state load balancers, and a QUIC extension exists {{QUIC_LB}} to
```
Also we have an own section on LB in this draft that we should probably refer to (instead). Maybe we can also just more some of the background infos in this section over there and then have both section right after each other...?!

mirjak reviewed (changes requested):
I would like to have this section a bit more crisp. There is probably some intro text that we don't need in this extent. But more importantly, things to do or not to do should be more clearly stated (as Brian also indicated). Maybe some text can also be moved in the load balancer section; plus double check with other text in this doc, e.g. in the sections on CID and Flow Associations.

martinduke commented on draft-ietf-quic-manageability.md:
```diff
@@ -905,6 +905,119 @@ need to be flow based.  However, 5-tuple (plus eventually connection ID if
+
+### The Lure of the Connection ID
+
+There are a few reasons that CID-aware NATs could seemingly appear attractive.
```
This text is gone

martinduke commented on draft-ietf-quic-manageability.md:
```diff
@@ -905,6 +905,119 @@ need to be flow based.  However, 5-tuple (plus eventually connection ID if
+outright, HTTP/3 over QUIC falls back to TCP. This is preferable to a
+connection suddenly black holing and timing out. Furthermore, wide deployment
+of NATs with this behavior would make it risky to change Connection IDs in the
+internet, which would thwart various important protocol properties.
```
Do you have a recommendation?

mirjak replied:
Maybe s/"make it risky to change Connection IDs"/"hinders the use of QUIC's migration function which relies on the ability to change the connection ID any time during the life time of a QUIC connection"...

martinduke commented on draft-ietf-quic-manageability.md:
```diff
@@ -905,6 +905,119 @@ need to be flow based.  However, 5-tuple (plus eventually connection ID if
+infrastructure that are almost never valid in client/NAT architectures.
+Deploying such a system would include the administrative overhead while not
+solving the problem described in this section if the client changes
+networks.
```
Do you have a suggestion?

mirjak replied:
I believe this belonged to the "assumptions about endpoint mobility and common configuration in server
 infrastructure" part. Not fully clear what assumptions you are talking about,

martinduke commented on draft-ietf-quic-manageability.md:
```diff
@@ -905,6 +905,119 @@ need to be flow based.  However, 5-tuple (plus eventually connection ID if
+solving the problem described in this section if the client changes
+networks.
+
+Note that using connection IDs in this manner would anyway violate the best
```
Port

martinduke commented on draft-ietf-quic-manageability.md:
```diff
@@ -905,6 +905,119 @@ need to be flow based.  However, 5-tuple (plus eventually connection ID if
+{{QUIC-TRANSPORT}}). The QUIC PATH_CHALLENGE and PATH_RESPONSE frames are
+intended to detect and mitigate these attacks and verify connectivity to the
+new address. This mechanism cannot work if the NAT is bleaching peer address
+changes.
```
I added another paragraph up front.

martinduke commented:
I addressed all of the comments except where I had no idea for what to write.

As for making the section more "crisp", I would much prefer that we land this PR and have another one composed by someone else who has a firm idea of what to do about it.

mirjak commented on draft-ietf-quic-manageability.md:
```diff
@@ -905,6 +905,117 @@ need to be flow based.  However, 5-tuple (plus eventually connection ID if
+While mid-connection failure in some cases may seem superior to rejecting QUIC
+outright, HTTP/3 over QUIC falls back to TCP, and
+{{?I-D.ietf-quic-applicability}} recommends TCP fallback for other protocols
+that use QUIC . This is preferable to a connection suddenly black holing and
```
```suggestion
that use QUIC, this is preferable to a connection suddenly black holing and
```

mirjak reviewed (approved):
Happy to merge for now and make further changes if needed later.

martinthomson commented on draft-ietf-quic-manageability.md:
```diff
@@ -905,6 +905,115 @@ need to be flow based.  However, 5-tuple (plus eventually connection ID if
+secretly changes its connection ID, there will be no mapping for the NAT, and
+the connection will suddenly break.
+
+While mid-connection failure in some cases may seem superior to rejecting QUIC
+outright, HTTP/3 over QUIC falls back to TCP, and
+{{?I-D.ietf-quic-applicability}} recommends TCP fallback for other protocols
+that use QUIC, this is preferable to a connection suddenly black holing and
+timing out, as the recovery time is much longer for a black hole event.
```
```suggestion
Some applications might prefer to risk mid-connection failure rather than
not being able to use QUIC. HTTP/3 prefers reliability, recommending that
TCP be used as a fallback and {{?I-D.ietf-quic-applicability}} recommends
TCP fallback for other protocols on the basis that this is preferable to
sudden connection errors and time outs.
```

I had trouble with the phrasing here and I didn't like the comma-splice much.

mirjak replied:
Didn't this completely change the meaning of the first sentence? I don't think it's true that any application would prefer mid-connection failure. The original sentence said "seem superior", meaning that it is not.

martinthomson replied:
"seems" is not a good way to phrase a negative; I read it in the positive.  I was trying for a more or less direct translation of the intent.  If you ask me, I think that you could be more assertive about this and say that 

> QUIC is deliberately designed to fail rather than persist when the network cannot support its operation.  For HTTP/3, this extends to recommending a fallback to TCP-based versions of HTTP rather than persisting with a QUIC connection that might be unstable.

britram reviewed (approved):
lgtm now thanks!
